File: backend/src/main/webapp/frontend/ts/picture/Picture.ts
import {FigureFactory} from "./figure/FigureFactory.js";
import {Figure} from "./figure/Figure.js";

export class Picture {
    constructor(
        public readonly id: Picture.Id,
        public readonly figure: Figure
    ) {
    }
}

export namespace Picture {
    export class Id {
        constructor(
            public readonly owner: string,
            public readonly name: string,
            public readonly version: string
        ) {
        }
    }

    export function fromJson(json: Map<string, any>): Picture {
        return new Picture(
            new Id(
                json.get('owner'),
                json.get('name'),
                json.get('version'),
            ),
            FigureFactory.fromJson(json.get('figure'))
        )
    }
}

File: backend/src/main/java/ru/vityaman/itmo/web/lab/servlets/backend/model/Picture.java
package ru.vityaman.itmo.web.lab.servlets.backend.model;

import lombok.Value;
import ru.vityaman.itmo.web.lab.servlets.backend.model.figure.Figure;

@Value
public final class Picture {
    private final Id id;
    private final Figure figure;

    public static Picture create(String ownerId, String name, Figure figure) {
        return new Picture(
            new Id(
                new User(ownerId),
                name,
                new Version(
                    Hashable.hash(String.format(
                        "%s:%s:%s",
                        ownerId, name, figure.hash()
                    ))
                )
            ),
            figure
        );
    }

    public static final class Version {
        private final String hash;

        public Version(String hash) {
            this.hash = hash;
        }

        public String hash() {
            return hash;
        }
    }

    @Value
    public static final class Id {
        private final User owner;
        private final String name;
        private final Version version;
    }
}
